File: src/main/java/com/zb/exam/prepare/GridBagLayoutDemo.java
package com.zb.exam.prepare;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class GridBagLayoutDemo {

    public static void main(String[] args) {
        JFrame frame = new JFrame("测试窗口");

        GridBagLayout gridBagLayout = new GridBagLayout();    // 布局管理器
        frame.setLayout(gridBagLayout);

        final JTextField tf1 = new JTextField();
        final JTextField tf2 = new JTextField();
        final JTextField tf3 = new JTextField();

        final JButton bt1 = new JButton("Start");
        final JButton bt2 = new JButton("Clear");


        GridBagConstraints c = null;                    // 约束

        c = new GridBagConstraints(0, 0, 1, 1, 0, 0,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL, new Insets(0, 0, 0, 0), 0, 0);
        frame.add(tf1);
        gridBagLayout.setConstraints(tf1, c);

        c = new GridBagConstraints(1, 0, 1, 1, 1, 0,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL, new Insets(0, 0, 0, 0), 0, 0);
        frame.add(tf2);
        gridBagLayout.setConstraints(tf2, c);

        c = new GridBagConstraints(0, 1, 2, 1, 0, 0,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL, new Insets(0, 0, 0, 0), 0, 0);
        frame.add(tf3);
        gridBagLayout.setConstraints(tf3, c);


        c = new GridBagConstraints(0, 2, 1, 1, 1, 0,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL, new Insets(0, 0, 0, 0), 0, 0);
        frame.add(bt1);
        gridBagLayout.setConstraints(bt1, c);
        bt1.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                if (e.getSource() == bt1) {
                    String text = tf1.getText() + tf2.getText();
                    tf3.setText(text);
                }
            }
        });


        c = new GridBagConstraints(1, 2, 1, 1, 1, 0,
                GridBagConstraints.CENTER, GridBagConstraints.HORIZONTAL, new Insets(0, 0, 0, 0), 0, 0);
        frame.add(bt2);
        gridBagLayout.setConstraints(bt2, c);
        bt2.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                if (e.getSource() == bt2) {
                    tf3.setText(null);
                }
            }
        });


        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        frame.setBounds(200, 200, 400, 400);
        frame.setVisible(true);
    }
}

File: src/main/java/com/zb/layout/GridBagLayoutDemo.kt
package com.zb.layout

import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JTextField

fun main() {
    val frame = JFrame("拨号盘")
    val layout = GridBagLayout()
    frame.layout = layout
    val tfConstraints = GridBagConstraints().apply {
        fill = GridBagConstraints.BOTH
        gridwidth = GridBagConstraints.REMAINDER
    }

    val tf = JTextField("123213123123").apply {
        layout.setConstraints(this, tfConstraints)
        frame.add(this)
    }

    val btConstraints = GridBagConstraints().apply {
        fill = GridBagConstraints.BOTH
        weightx = 5.0
        weighty = 2.0
    }
    val btEndConstraints = GridBagConstraints().apply {
        fill = GridBagConstraints.BOTH
        gridwidth = GridBagConstraints.REMAINDER
        weightx = 5.0
        weighty = 2.0
    }

    frame.addButton("7", btConstraints)
    frame.addButton("8", btConstraints)
    frame.addButton("9", btEndConstraints)
    frame.addButton("4", btConstraints)
    frame.addButton("5", btConstraints)
    frame.addButton("6", btEndConstraints)
    frame.addButton("1", btConstraints)
    frame.addButton("2", btConstraints)
    frame.addButton("3", btEndConstraints)
    frame.addButton("返回", btConstraints)
    frame.addButton("拨号", btEndConstraints)

    frame.setBounds(200, 200, 400, 400)
    frame.isVisible = true
    frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
}

fun JFrame.addButton(title: String, constraints: GridBagConstraints): JButton {
    val bt = JButton(title)
    (this.contentPane.layout as GridBagLayout).setConstraints(bt, constraints)
    add(bt)
    return bt
}
